[FIX] project: show subtasks in project if converted into tasks

hadielyakhni opened the pull request:
Steps to reproduce:
1. Create a task, and a subtask
3. Open the subtask
4. From the gear icon, click "Convert to Task/Sub-Task"
5. Remove the parent task and click "Convert Task"
6. Go to the project dashboard > the subtask is not shown.

Why?
The above reproducing steps lead to the subtask having the `display_in_project` stored as `False`, hence it's not visible on the dashboard page. This is because `_compute_project_id` interrupted the computation flow of `display_in_project` by a call to `remove_to_compute` [1].

Fix:
In the `write` method, explicitly set `display_in_project` to `True` if the parent_id has been removed.

[1]: https://github.com/odoo/odoo/blob/9dec8329523d4966243c6491e7e556e643af6cdf/addons/project/models/project_task.py#L337

opw-4384222

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/191530.png)](https://mergebot.odoo.com/odoo/odoo/pull/191530)

taqu-odoo commented:
Makes sense to me.
Another possibility would be to add the `display_in_project` as an invisible field to the `project_task_convert_to_subtask_view_form`, so it is computed and saved when the "Convert to task" form is closed. But as this fix will work on existing databases it seems better.

hadielyakhni commented:
Thank you for the suggestion @taqu-odoo 👌🏼 

I might have missed something, but I have tried your suggestion, and the field didn't get computed when closing the modal.

xavierbol commented:
We could add `display_in_project = True` in the vals given in write method if `vals['parent_id'] == False`. It would be easier, IMO.

What do you think?

hadielyakhni commented:
Hello @xavierbol @auon-odoo,
I have tried your suggestion, but it's not working 👀 
```py
if 'parent_id' in vals and vals.get('parent_id') == False:
    vals['display_in_project'] = True
```

xavierbol commented:
Hello @hadielyakhni 

Why is it not working? In which case? 🧐

hadielyakhni commented:
@xavierbol, oh so if we applied your suggestion and reverted mine, somehow, the test I wrote in the PR fails; I also tried the flow myself and the subtask didn't get visible in the project dashboard which matches with the test failing.

xavierbol commented:
Hello @hadielyakhni 

I still don't understand why it does not work.

I quickly tested by doing that: 

```diff
diff --git a/addons/project/models/project_task.py b/addons/project/models/project_task.py
index 48baa7bbe546..6bd53ef9a133 100644
--- a/addons/project/models/project_task.py
+++ b/addons/project/models/project_task.py
@@ -1156,6 +1156,8 @@ class Task(models.Model):
         return tasks

     def write(self, vals):
+        if vals.get('parent_id') is False:
+            vals['display_in_project'] = True
         if len(self) == 1:
             handle_history_divergence(self, 'description', vals)
         portal_can_write = False
```

and seems good enough to fix the issue.

May I miss something?

xavierbol commented on addons/project/tests/test_project_subtask_ui_flow.py:
```diff
@@ -0,0 +1,22 @@
+            'project_id': self.project_goats.id,
+            'child_ids': [Command.create({ 'name': 'Sub-task invisible', 'project_id': self.project_goats.id, })],
+        })
+        invisible_subtask, = task.child_ids
```
Since there is only one subtask, we could do that:
```suggestion
        invisible_subtask = task.child_ids
```

xavierbol commented on addons/project/tests/test_project_subtask_ui_flow.py:
```diff
@@ -0,0 +1,22 @@
+        task = self.env['project.task'].create({
+            'name': 'Parent task',
+            'project_id': self.project_goats.id,
+            'child_ids': [Command.create({ 'name': 'Sub-task invisible', 'project_id': self.project_goats.id, })],
```
```suggestion
            'child_ids': [Command.create({'name': 'Sub-task invisible', 'project_id': self.project_goats.id})],
```

xavierbol commented on addons/project/tests/test_project_subtask_ui_flow.py:
```diff
@@ -0,0 +1,22 @@
+
+        with Form(invisible_subtask, view="project.project_task_convert_to_subtask_view_form") as subtask_form:
+            subtask_form.parent_id = self.env['project.task']
+            invisible_subtask = subtask_form.save()
```
No need, when we leave the `with` block, the form will be automatically saved.

```suggestion
```

xavierbol commented on addons/project/tests/test_project_subtask_ui_flow.py:
```diff
@@ -0,0 +1,22 @@
+from .test_project_base import TestProjectCommon
+
+@tagged('-at_install', 'post_install')
+class TestProjectSubtaskUIFlow(HttpCase, TestProjectCommon):
```
Why do we need `HttpCase`? 🤔 

Also, I think you can move your test in `test_project_subtasks` python file

hadielyakhni replied:
`HttpCase` was not needed, thanks for pointing that out :)

xavierbol reviewed (commented):
Thanks for your work, I left some additional comments.

hadielyakhni commented:
@xavierbol Thanks for the review and the suggestions.

> I still don't understand why it does not work.

My bad on this, indeed your suggestion works (It didn't worked before because I have added it in the wrong position 😬).
I updated the PR following your suggestions :)

I'll ping you again when it's green :) thank you.

xavierbol reviewed (approved):
robodoo r+
